### PR TITLE
[Pboro/Oxon] Move assets to configuration.

### PR DESCRIPTION
**Pull request by dracos**
[skip changelog] Bucks have asked (FD-2580) for some new staff categories to be linked to assets, so thought might as well move them first to save having to do that in the code.

<details>
<summary>New Pboro asset configuration</summary>
<pre>
peterborough:
- - name: 'default'
    max_resolution: 4.777314267158508
    srsName: "EPSG:3857"
    body: "Peterborough City Council"
  - name: 'tilma'
    template: 'default'
    http_wfs_url: "!!(* $tilma_host *)!!/mapserver/peterborough"
    geometryName: 'msGeometry'
  - name: 'arcgis'
    template: 'default'
    protocol_class: OpenLayers.Protocol.ArcgisHTTP
    format_class: OpenLayers.Format.GeoJSON
    http_options:
      params:
        inSR: '3857'
        outSR: '3857'
        f: 'geojson'
    geometryName: 'SHAPE'
  - name: 'trees'
    template: 'tilma'
    class: OpenLayers.Layer.VectorAssetMove
    select_action: true
    actions:
      asset_found: fixmystreet.message_controller.asset_found
      asset_not_found: fixmystreet.message_controller.asset_not_found
    attributes:
      tree_code: 'TREE_CODE'
    asset_id_field: 'TREE_CODE'
    asset_group: 'Trees'
    relevant: fixmystreet.assets.peterborough.trees_relevant
  - name: 'lights'
    template: 'tilma'
    wfs_feature: "StreetLights"
    asset_id_field: 'UNITID'
    asset_type: 'spot'
    asset_item: 'light'
  - name: 'bin'
    template: 'arcgis'
    http_options:
      url: 'https://tilma.mysociety.org/resource-proxy/proxy.php?https://peterborough.assets/0/query?'
      params:
        inSR: '3857'
        outSR: '3857'
        f: 'geojson'
        outFields: 'OBJECTID,Reference,Location'
    class: OpenLayers.Layer.VectorAssetMove
    select_action: true
    actions:
      asset_found: fixmystreet.message_controller.asset_found
      asset_not_found: fixmystreet.message_controller.asset_not_found
    attributes:
      asset_details: fixmystreet.assets.peterborough.bin_asset_details
      central_asset_id: 'OBJECTID'
    asset_id_field: 'OBJECTID'
    asset_type: 'spot'
  - name: 'flytipping'
    template: 'arcgis'
    http_options:
      params:
        inSR: '3857'
        outSR: '3857'
        f: 'geojson'
        outFields: ''
    # this prevents issues when public and non public land are right next to each other
    nearest_radius: 0.01
    stylemap: fixmystreet.assets.stylemap_invisible
    asset_category: ['General fly tipping', 'Hazardous fly tipping', 'Offensive graffiti', 'Non offensive graffiti', 'Offensive graffiti - STAFF ONLY' ]
    non_interactive: true
    road: true
    asset_item: 'road'
    asset_type: 'road'
- template: 'arcgis'
  http_options:
    url: 'https://tilma.mysociety.org/resource-proxy/proxy.php?https://peterborough.assets/7/query?'
    params:
      outFields: 'USRN'
  nearest_radius: 2
  stylemap: fixmystreet.assets.stylemap_invisible
  non_interactive: true
  always_visible: true
  usrn:
    attribute: 'USRN'
    field: 'site_code'
  name: "Adopted Highways"
- template: 'trees'
  wfs_feature: "tree_groups"
  asset_type: 'area'
  asset_item: 'tree group'
- template: 'trees'
  wfs_feature: "tree_points"
  asset_type: 'spot'
  asset_item: 'tree'
# We don't want to plant trees where the existing trees are, so add a
# separate layer with pin-snapping disabled for new tree requests.
# The new tree request category is disabled in the other tree point layer.
- template: 'tilma'
  wfs_feature: "tree_points"
  asset_id_field: 'TREE_CODE'
  asset_type: 'spot'
  asset_category: 'Request for tree to be planted'
  asset_item: 'tree'
  disable_pin_snapping: true
  asset_item_message: ''
- template: 'light'
  class: OpenLayers.Layer.VectorAssetMove
  stylemap: fixmystreet.assets.peterborough.streetlight_stylemap
  feature_code: 'UNITNO'
  attributes:
    asset_details: fixmystreet.assets.peterborough.lighting_asset_details
  asset_group: 'Street lighting'
  asset_category: 'Lighting enquiry'
  relevant: fixmystreet.assets.peterborough.lighting_relevant
  select_action: true
  actions:
    asset_found: fixmystreet.assets.peterborough.lighting_asset_found
    asset_not_found: fixmystreet.assets.peterborough.lighting_asset_not_found
- template: 'light'
  asset_category: 'Problem with a light not shown on map'
  disable_pin_snapping: true
  asset_item_message: ''
- template: 'bin'
  http_options:
    params:
      where: "Type='Litter Bin'"
  asset_category: 'Litter bin'
  asset_item: 'litter bin'
- template: 'bin'
  http_options:
    params:
      where: "Type='Dog Waste Bin'"
  asset_category: 'Dog bin'
  asset_item: 'dog waste bin'
# PCC Property Combined
- template: 'flytipping'
  http_options:
    url: 'https://tilma.mysociety.org/resource-proxy/proxy.php?https://peterborough.assets/4/query?'
  actions:
    found: fixmystreet.assets.peterborough.flytipping_pcc_found
    not_found: fixmystreet.assets.peterborough.flytipping_pcc_not_found
# PCC Property Leased Out NOT Responsible
- template: 'flytipping'
  http_options:
    url: 'https://tilma.mysociety.org/resource-proxy/proxy.php?https://peterborough.assets/3/query?'
  actions:
    found: fixmystreet.assets.peterborough.flytipping_leased_found
</pre>
</details>

<details>
<summary>New Oxfordshire asset configuration</summary>
<pre>
    oxfordshire:
      - - name: 'default'
          wfs_url: "!!(* $tilma_host *)!!/mapserver/oxfordshire"
          asset_type: 'spot'
          max_resolution: 4.777314267158508
          geometryName: 'msGeometry'
          srsName: "EPSG:3857"
          body: "Oxfordshire County Council"
        - name: 'owned_base'
          template: 'default'
          select_action: true
          srsName: "EPSG:27700"
          actions:
            asset_found: fixmystreet.assets.oxfordshire.owned_asset_found
            # Not a typo, asset selection is not mandatory
            asset_not_found: fixmystreet.message_controller.asset_found
        - name: 'owned_defaults'
          template: 'owned_base'
          stylemap: fixmystreet.assets.oxfordshire.owned_stylemap
          # have to do this by hand rather than using wfs_* options
          # as the server does not like being POSTed xml with application/xml
          # as the Content-Type which is what using those options results in.
          http_options:
            headers:
              'Content-Type': 'text/plain'
            params:
              SERVICE: "WFS"
              VERSION: "1.1.0"
              REQUEST: "GetFeature"
              SRSNAME: "urn:ogc:def:crs:EPSG::27700"
              propertyName: 'id,maintained_by,msGeometry'
          asset_id_field: 'id'
          attributes:
            feature_id: 'id'
          owns_function: fixmystreet.assets.oxfordshire.owns_feature
        - name: 'drain_defaults'
          template: 'owned_defaults'
          http_options:
            url: "!!(* $tilma_host *)!!/proxy/occ/drains/wfs"
          asset_category: ['Drainage', "Gully and Catchpits"]
        - name: 'drain_inspection_defaults'
          template: 'drain_defaults'
          stylemap: fixmystreet.assets.stylemap_invisible
          non_interactive: true
        - name: 'drain_asset_defaults'
          template: 'drain_defaults'
          http_options:
            params:
              propertyName: 'id,msGeometry,maintained_by,asset_id,created,last_inspected'
          select_action: true
          construct_selected_asset_message: fixmystreet.assets.oxfordshire.drain_construct_selected_asset_message
          actions:
            asset_found: fixmystreet.assets.named_select_action_found
            asset_not_found: fixmystreet.assets.named_select_action_not_found
        - name: 'oxfordshire_defaults'
          format_class: OpenLayers.Format.GeoJSON
          srsName: "EPSG:4326"
          class: OpenLayers.Layer.VectorAssetMove
          non_interactive: false
          body: "Oxfordshire County Council"
          attributes:
            # feature_id
            unit_number: "title"
            unit_type: "unit_type"
            asset_resource_id: "itemId"
          select_action: true
          feature_code: 'title'
          asset_id_field: 'itemId'
          construct_selected_asset_message: fixmystreet.assets.oxfordshire.light_construct_selected_asset_message
          actions:
            asset_found: fixmystreet.assets.oxfordshire.light_asset_found
            asset_not_found: fixmystreet.assets.oxfordshire.light_asset_not_found
      - template: 'default'
        stylemap: fixmystreet.assets.oxfordshire.stylemap
        wfs_feature: "Trees"
        asset_id_field: 'Ref'
        attributes:
          feature_id: 'Ref'
        asset_category: ["Trees"]
        asset_item: 'tree'
      - template: 'default'
        select_action: true
        stylemap: fixmystreet.assets.oxfordshire.stylemap
        wfs_feature: "Traffic_Lights"
        asset_id_field: 'Site'
        attributes:
          feature_id: 'Site'
        asset_category: ["Traffic Lights (permanent only)"]
        asset_item: 'traffic light'
        feature_code: 'Site'
        actions:
          asset_found: fixmystreet.assets.named_select_action_found
          asset_not_found: fixmystreet.assets.named_select_action_not_found
      - template: 'drain_inspection_defaults'
        http_options:
          params:
            propertyName: 'id,msGeometry,asset_id,created,junction_cleaned'
            TYPENAME: "junction_inspections"
        asset_item: 'drain'
        name: "Oxon Junction Inspections"
      - template: 'drain_inspection_defaults'
        http_options:
          params:
            propertyName: 'id,msGeometry,asset_id,created,channel_cleaned'
            TYPENAME: "channel_inspections"
        asset_item: 'gully'
        name: "Oxon Channel Inspections"
      - template: 'drain_asset_defaults'
        http_options: { params: { TYPENAME: "channels" } }
        asset_item: 'gully'
      - template: 'drain_asset_defaults'
        http_options: { params: { TYPENAME: "junctions" } }
        asset_item: 'drain'
      - template: 'owned_defaults'
        http_options:
          url: "!!(* $tilma_host *)!!/proxy/occ/grit/wfs"
          params:
            TYPENAME: "Grit_bins"
        asset_category: ["Ice/Snow"]
        asset_item: 'grit bin'
      - template: 'owned_base'
        stylemap: fixmystreet.assets.oxfordshire.bridge_stylemap
        wfs_url: "!!(* $tilma_host *)!!/proxy/occ/nsg/"
        wfs_feature: 'BRIDESFMS1'
        geometryName: 'SHAPE_GEOMETRY'
        propertyNames: ['ALL_STRUCTURES_CODE', 'MAINTENANCE_AUTHORITY_UID', 'SHAPE_GEOMETRY']
        filter_key: 'MAINTENANCE_AUTHORITY_UID'
        filter_value: [1, 21]
        asset_category: ['Bridges']
        asset_item: 'bridge'
        asset_id_field: 'ALL_STRUCTURES_CODE'
        attributes:
          feature_id: 'ALL_STRUCTURES_CODE'
        no_asset_msg_id: '#js-occ-prow-bridge'
        owns_function: fixmystreet.assets.oxfordshire.owns_bridge
      - template: 'default'
        stylemap: fixmystreet.assets.stylemap_invisible
        wfs_url: "!!(* $tilma_host *)!!/proxy/occ/nsg/"
        wfs_feature: 'MAINTAINABLE_AT_PUBLIC_EXPENSE_FMS'
        geometryName: 'SHAPE_GEOMETRY'
        propertyNames: ['UNIQUE_STREET_REFERENCE_NUMBER', 'SHAPE_GEOMETRY']
        srsName: "EPSG:27700"
        usrn:
          attribute: 'UNIQUE_STREET_REFERENCE_NUMBER'
          field: 'usrn'
        non_interactive: true
        road: true
        no_asset_msg_id: '#js-not-a-road'
        asset_item: 'road'
        asset_type: 'road'
        actions:
          found: fixmystreet.message_controller.road_found
          not_found: fixmystreet.message_controller.road_not_found
        asset_category:
          - "Cover Hanging"
          - "Door Missing"
          - "Flashing Lamp"
          - "Vehicle/ Accident Damage"
          - "Knocked Down Bollard"
          - "Lamp Appears Dim"
          - "Lamp On During Day"
          - "Lamp Out of Light"
          - "Other Lighting Issue"
          - "Twisted Lantern"
          - "Bridges"
          - "Carriageway Defect"
          - "Current Roadworks"
          - "Drainage"
          - "Gully and Catchpits"
          - "Highway Schemes"
          - "Ice/Snow"
          - "Manhole"
          - "Pavements"
          - "Pothole"
          - "Road Traffic Signs and Road Markings"
          - "Roads/highways"
          - "Traffic Lights (permanent only)"
          - "Trees"
      - template: 'oxfordshire_defaults'
        http_options: { url: "!!(* $tilma_host *)!!/alloy/oxfordshire-lights.php?url=!!(* $alloy_url *)!!" }
        stylemap: fixmystreet.assets.oxfordshire.streetlight_stylemap
        asset_category:
          - "Cover Hanging"
          - "Door Missing"
          - "Flashing Lamp"
          - "Vehicle/ Accident Damage"
          - "Knocked Down Bollard"
          - "Lamp Appears Dim"
          - "Lamp On During Day"
          - "Lamp Out of Light"
          - "Other Lighting Issue"
          - "Twisted Lantern"
        max_resolution: 1.194328566789627
        asset_item: "street light"
        asset_type: "spot"
</pre>
</details>

### CODE REVIEW

**Review by davea — CHANGES_REQUESTED** [review]
Looks good, but a couple of small issues with the `asset_item_message` values in the suggested config:

The `»` character isn't encoded correctly:
image

Also the suggested YAML doesn't parse due to the linebreaks in the two `asset_item_message` values. I tried putting `\n` literals in, which just show up as `\n` in the message:
image

Using a multiline value parses OK but the linebreaks seem to be ignored/stripped:

```yaml
        asset_item_message: |
          You can pick a ITEM from the map »
          If there is no yellow dot showing then this ITEM is not maintained by Transport for Buckinghamshire
```
image

File: data/test-asset-layers.yml [hunk 1]
```diff
@@ -16,7 +16,7 @@ bathnes:
     asset_type: 'spot'
     max_resolution: 4.777314267158508
     asset_id_field: 'feature_no'
-    attributes: null
+    attributes: ~
```
[the remark this answers is not in the document]
Comment (dracos) [reply data/test-asset-layers.yml:19]: It seems that null in our parser is becoming the string 'null', yes.

**Comment by dracos** [comment]
"The » character isn't encoded correctly:" - same issue I just fixed on Pboro bulky goods, perhaps, will look.

"Also the suggested YAML doesn't parse due to the linebreaks in the two asset_item_message values." - this is the GitHub markdown text changing the br that is there into an actual line break, sorry, didn't realise it had done that. It is `<br>` in the actual config :)

**Comment by davea** [comment]
> this is the GitHub markdown text changing the br

🤦

**Comment by dracos** [comment]
Encoding was the same issue, have included fix for that, and also as I'd halfway done it last week, have also added Peterborough as another move.

**Review by davea — APPROVED** [review]
:shipit:

**Review by davea — APPROVED** [review]
💡

**Pull request merged**